#376 - Add logging standards from MBTP (#377)
UKHomeOffice/engineering-guidance-and-standards: docs/standards/logging.md

---
layout: standard
order: 1
title: Logging for services with centralised logging aggregation
date: 2024-01-24
id: SEGAS-00015 
tags:
  - SRE
  - Monitoring
  - Observability
  - Logging
related: 
  sections:
    - title: Securing application logging
      items:
        - text: Securing application logging
          href: /patterns/securing-application-logging/ 
---

> Logging tools are one of the primary tools to debug application problems and identifying the root cause of incidents. However, logs need to be formatted in a standard way so that they are not only useful for the application's original developers, but also those supporting the application later in its life-cycle. 

Reviewer: ```suggestion
Logging tools are one of the primary tools to debug application problems and identifying the root cause of incidents. Logs need to be formatted in a standard way so that they are useful for the both application's developers and those supporting the application later in its life-cycle. 
```

> However, logs are sometimes relied on too heavily for problem detection, when this role would be better suited to purpose chosen monitoring tooling. 

Reviewer 2: ```suggestion
However, logs are sometimes relied on too heavily for problem detection, when this role would be better suited to monitoring tooling. 
```

> However, logs are sometimes relied on too heavily for problem detection, when this role would be better suited to purpose chosen monitoring tooling. 
> 
> Therefore these standards provide guidance on both the proper format of application logs, as well as their proper usage.

Reviewer: ```suggestion
Logs are not a replacement for dedicated monitoring tooling. They should be used as part of a broader observabilty toolset. This standard provides guidance on a standard format for application logs, and suitable uses.
```

> Therefore these standards provide guidance on both the proper format of application logs, as well as their proper usage.

Reviewer 2: ```suggestion
This standard provides guidance on the proper format and usage of logs.
```

- [Service logs MUST have an identifiable source in the log aggregator](#service-logs-must-have-an-identifiable-source-in-the-log-aggregator)
- [Service log messages should only be used for metrics, dashboarding and alerting as a last resort](#service-log-messages-should-only-be-used-for-metrics%2C-dashboarding-and-alerting-as-a-last-resort)
- [There MUST be no DEBUG / TRACE level messages in Production environments](#there-must-be-no-debug-%2F-trace-level-messages-in-production-environments)
- [Service log messages MUST not reveal any sensitive information or person identifiable information](#service-log-messages-must-not-reveal-any-sensitive-information-or-person-identifiable-information)

### Service logs MUST have the minimum required fields

Information | Description
------------| -----------
 Date / Time | Each log message must include date and time (millisecond accuracy) (UTC). The time source for all service logs must be consistent so that messages can be viewed & sorted easily.
Service | Each log message must be attributed to an service (i.e. hostname).
Log level | Each log message must have an appropriate log level (INFO, WARN, ERROR, etc.)

> Log Message | Each log message must have a log message. Log messages must follow a consistent log messaging format for all log types to allow for easy processing and translation of logs to other formats.

Reviewer 2: Tables can cause flow issues we have found. Could consider using summary lists as in the [docs-as-code pattern](https://github.com/UKHomeOffice/engineering-guidance-and-standards/blob/main/docs/patterns/docs-as-code.md?plain=1)
